go/store/datas/database_common.go

```diff
@@ -417,6 +417,78 @@ func (db *database) doCommit(ctx context.Context, datasetID string, commit types
 	return tryCommitErr
 }
 
+func (db *database) Tag(ctx context.Context, ds Dataset, ref types.Ref, opts TagOptions) (Dataset, error) {
+	return db.doHeadUpdate(
+		ctx,
+		ds,
+		func(ds Dataset) error {
+			st, err := NewTag(ctx, ref, opts.Meta)
+
+			if err != nil {
+				return err
+			}
+
+			return db.doTag(ctx, ds.ID(), st)
+		},
+	)
+}
+
+// doTag manages concurrent access the single logical piece of mutable state: the current Root. It uses
+// the same optimistic writing algorithm as doCommit (see above).
+func (db *database) doTag(ctx context.Context, datasetID string, tag types.Struct) error {
+	if is, err := db.validateTag(ctx, tag); err != nil {
+		return err
+	} else if !is {
+		d.Panic("Can't tag using a non-Commit struct (dataset %s)", datasetID)
+	}
+
+	// This could loop forever, given enough simultaneous writers. BUG 2565
+	var tryCommitErr error
+	for tryCommitErr = ErrOptimisticLockFailed; tryCommitErr == ErrOptimisticLockFailed; {
+		currentRootHash, err := db.rt.Root(ctx)
+
+		if err != nil {
+			return err
+		}
+
+		currentDatasets, err := db.Datasets(ctx)
+
+		if err != nil {
+			return err
+		}
+
+		tagRef, err := db.WriteValue(ctx, tag) // will be orphaned if the tryCommitChunks() below fails
+
+		if err != nil {
+			return err
+		}
+
+		_, hasHead, err := currentDatasets.MaybeGet(ctx, types.String(datasetID))
+
+		if err != nil {
+			return err
+		}
+
+		if hasHead {
+			d.Panic("datasets for tags (refs/tags/*) cannot be altered after creation")
```

Comment:
Should this be an `err` instead? Seems like it might represent a race condition, and not some fatal programming error in dolt itself.